File: src/core/refs.ts
import { extractRefs } from "./interpolate.ts";

// Who references ${target}? Used by var/global remove, unwire and vault remove
// to build DEPENDENT_REFERENCE blockers (spec: "Dependency detection"). The
// caller assembles records from every vault it can open; vaults it cannot open
// become UNVERIFIED_REFERENCES warnings at the planning layer.
export interface ValueRecord {
  variable: string;
  vault: string;
  consumer: string;
  raw: string;
}

export interface Dependent {
  variable: string;
  vault: string;
  consumer: string;
}

export function findDependents(target: string, records: Iterable<ValueRecord>): Dependent[] {
  const out: Dependent[] = [];
  for (const r of records) {
    if (extractRefs(r.raw).includes(target)) {
      out.push({ variable: r.variable, vault: r.vault, consumer: r.consumer });
    }
  }
  return out;
}

File: go/internal/core/refs.go
package core

// ValueRecord holds one raw vault value for dependency scanning.
type ValueRecord struct {
	Variable string
	Vault    string
	Consumer string
	Raw      string
}

// Dependent is a (variable, vault, consumer) triple that references a target.
type Dependent struct {
	Variable string
	Vault    string
	Consumer string
}

// FindDependents returns all records whose raw value contains a ${target} ref.
func FindDependents(target string, records []ValueRecord) []Dependent {
	var out []Dependent
	for _, r := range records {
		for _, ref := range ExtractRefs(r.Raw) {
			if ref == target {
				out = append(out, Dependent{
					Variable: r.Variable,
					Vault:    r.Vault,
					Consumer: r.Consumer,
				})
				break
			}
		}
	}
	return out
}
